src/types/block.zig

```diff
@@ -41,6 +45,28 @@ pub const BlockHeader = struct {
         allocator.free(self.extra_data);
         self.* = undefined;
     }
+
+    // defines a custom encoder so that fields that were added later on in the chain history
+    // are not included in the RLP when hashing.
+    pub fn encodeToRLP(self: *const BlockHeader, allocator: Allocator, list: *std.ArrayList(u8)) !void {
```

Comment:
I wonder if we can make optional fields behave this automatically in the default RLP encoder? Feels like the correct default behavior avoiding every case having to do this custom encoder manually. These if statements with all the repeated fields are quite hairy to spot if you missed one field or similar.

Reply:
I considered doing the same thing at first, but then realized the following points:

 * This is not a rlp issue, it's a protocol issue: rlp isn't designed to have optional fields. I don't think we should burden the  base lib with this concept that is mainly used for blocks (and once in some p2p subprotocol). Note that this is also not the approach that geth has taken - they use a custom encoder instead.
 * Even if we did, it means that once you run into a `null` optional, you need to figure out if things stop there. So you need to somehow backtrack, which makes for complicated code.

I just got an idea writing this, which might make this custom implementation more maintainable, so I'll hold off on the merging.